### ISSUE TITLE
Serverless Capabilities

### ISSUE THREAD

**Issue by tanandy** [ovh/public-cloud-roadmap#90]
As a user  I want to be able to use serverless capabilities (serverless functions, serverless databases ...) for the public cloud
I want to be able to consume these features through standard APIs / OVHcloud Customer panel

**Comment by mhurtrel** [comment]
@tanandy Managed Databases are indeed on the short term roadmap , do not hesitate to subscribe and vote on  https://github.com/ovh/public-cloud-roadmap/issues/25 and share in comment your priorities.

Concerning serverless function and serverless containers we are still evaluating options. We don't see a clear, stable and totally reversible technology that would make a consensus at this stage, but I am interested in you wharing more details about your needs in terms of features/API features and billing model

**Comment by tanandy** [comment]
ok for all serverless stuff billing, we expect pay as you go model for the resource we use (ex: go used , traffic ...)

for example, we expected to get some higher disk space for managed serverless db that scale automatically (ex: 128 TO max)

for functions its basically, beeing able to run some serverless architecture or just a basic task ...

for serverless containers, we need to be able to run easily containers without even managing it a lot (CaaS , PaaS)

**Comment by tanandy** [comment]
for the technology i think you could give a try to  https://www.openfaas.com/  and some french OVH concurrents made a consensus.

**Comment by tanandy** [comment]
@mhurtrel what about knative ?

**Comment by Moumouls** [comment]
@mhurtrel i agree with @tanandy , knative was recently transferred to CNCF 
https://www.cncf.io/blog/2022/03/02/knative-accepted-as-a-cncf-incubating-project/

Many developers are now familiar with kubenertes, I hope OVH will support knative soon.
I like to work with OVH since it's a french company.

But currently, in my company we plan to move out of OVH to gcp and their knative support on top of cloud run.

(Note: my company is subject to highly variable traffic, so we need a serverless context to avoid some future cloud billing issues, and scalability limitations)

So I hope the knative option will be checked ☺️

**Comment by mhurtrel** [comment]
@Moumouls @tanandy Indeed, we are currently observing some alignment on a couple of standards for this products, and will consider offering serverless capabilities in the mid term. I add this ticket to the backlog, open for comments.

Note that knative will be offering in our Hosted Private Cloud powered by Anthos in the short-term, so the scalability model (monthly commitment on baremetal nodes, that you can dynamically providion to a user cluster where you deployed knative serving and eventing) may not fit your use case with highly variable traffic @Moumouls 

I am very interested in feedbacks on your use cases for a container-as-a-service product like this  :
- order of magnitude of scalability needed (scale to 0 ? cold starts acceptable timings ?)
- localisation ?
- stateteless, with high speed connectivity to object storage would be sufficient ou management of perisstent volumes would be a requirement ?
- usage of competing technologies/products and the potential limits you hit

**Comment by mhurtrel** [comment]
@tanandy I edited the issue content and title to focus on CaaS/FaaS, excluding serverless databases, which are a quite different challenge/design. DO not hesitate to open a dedicated issue for this with detailed requirement, that would interest my colleague @baaastijn for sure

**Comment by Moumouls** [comment]
> order of magnitude of scalability needed (scale to 0 ? cold starts acceptable timings ?)

Scale to 0 sure. About cold starts, in Caas on your side i suppose that the main goal will be to optimize image caching/availability in the region. A really first cold start of a new image version don't matter. But once the image is already used by your servers, we can expect nearly instant container start (may be by optimizing container/request scheduling).
Another parameter about this. In Caas use case, and Knative, traffic is only routed to the container once the container is listening on the target port. So the cold start (if we don't scale to 0) only delay the scalability reactivity. And developers have many techniques to optimize the server listening, for example lazy module import in NodeJS.

Note: also to avoid many start & stop, a container should be stopped after may 15min without any requests. I do not expect to be charged for idle time of a container, or like GCP it could be a pricing option.

> localisation ?

If the container registry is global, a multi region could be nice. But i think also GCP provide a specific service url for each region since it allow developers to choose their favorite load balancing service (like cloudflare) on top.

> stateteless, with high speed connectivity to object storage would be sufficient ou management of perisstent volumes would be a requirement ?

From my point of view, a serverless app should not rely on persistent volume (volume attachment could take lot of time). GCP for example do not support this feature (for obvious reasons).
Not sure to understand the high speed connectivity to object storage. But yes the serverless service need to be deployed on many datacenters to be able to be close to other cloud services datacenters (from AWS, DO, GCP). For example, OVH is not heavily integrated in some 3rd party services (Mongo Atlas and their new Serverless feature for example), yes connectivity/localisation is important.


> usage of competing technologies/products and the potential limits you hit

Today the most complete Caas seems GCP to me.

Pro: 
- CPU billing based on request cpu usage. An idle cpu is charged much less.
- Knative interface on top of cloud run custom interface
- GA available so my containers can run in the same datacenter as for example the Mongo Serverless service
- IAM, identity support to be able to only accept request from specific sources (like CF)

Cons:
- Knative interface not supported for Cron jobs
- Pricing is really hard to predict if you do not have already stats on you infra usage/request/bandwidth

I currently do not checked how secrets are managed.

**Comment by mhurtrel** [comment]
@Moumouls  Thanks a lot for this prompt and detailed feedback !

**Comment by tanandy** [comment]
i will split this in different issue since FaaS is different from CaaS

**Comment by tanandy** [comment]
Basic requirements: 

- Pay as you go
- Automatically handles the deployment, from capacity provisioning, load balancing, auto-scaling to application health monitoring
- Scale to zero 
- Cold Starts: yes (to be defined)

**Comment by aehlke** [comment]
I used to run devops / continuous deployment transformation teams and absolutely agree with @tanandy 's requirements. I would however say that a huge market differentiator would be to have some kind of prepaid credit as well where you can have fair warning and a shutoff (with short term mitigation steps) rather than a huge bill. It's a fear that hangs over the heads of smaller/independent businesses

**Comment by jcrben** [comment]
@mhurtrel @aehlke echoing the sentiment about prepaid credits and auto shutoff

even at large enterprises, mistakes can do happen, so this would improve peace of mind

OVH already offers a system for prepaid credits so this could piggyback on that

**Comment by drzraf** [comment]
2021 -> 2026, CAAS ?

Maybe just enable OpenStack Zun on your OpenStack instance. The guide is here https://docs.openstack.org/zun/latest/install/index.html and shouldn't be _that_ difficult. Advert the service as "experimental", keep it cheap, let it scale, provide tight integration with GPUs.

After 5 years, it may be better to _just ship something_, even suboptimal, instead of postponing and missing entire opportunity windows (and clients needs).

**Comment by antonin-a** [comment]
Hi @drzraf, fair point, I won't argue with the dates: this issue has been open far too long. Assignee and PM for this product here, so let me be concrete:

Your timing is actually good, we are preparing an alpha of our Serverless platform (CaaS + FaaS), targeted for the end of this year. Deploy a container image (or source code) and get an HTTPS endpoint with autoscaling including scale-to-zero, pay-per-use billing, revisions and rollback, logs and metrics, which is what this thread has been asking for from the start.

On the Zun suggestion: we did look at lighter paths. The honest answer is that exposing a container runtime on shared OpenStack would not deliver the contract this thread consistently described (scale-to-zero, per-request billing, strong tenant isolation, a fully managed developer experience). Shipping that as "CaaS" would have closed the issue on paper but probably disapointing. So we built the real thing, and it is now close. (GPU-backed workloads are on the roadmap too, though not in the alpha.)

I'll post here when the alpha program opens so people on this thread can register as early testers. Meanwhile, if you want to influence what lands first you can already share your top use cases (runtime, triggers, scale profile). This thread has directly shaped the requirements more than once.

**Comment by davidaparicio** [comment]
While waiting for a dedicated Serverless / Container-as-a-Service offering on Public Cloud, I'm currently using OVHcloud AI Deploy as a workaround to get part of this functionality back: managed container deployment, automatic scaling, exposure through a standard API, and management via the OVHcloud Manager or the CLI with ```ovhai```.

This covers part of the need described in this ticket (deploying containers without managing the underlying infrastructure), but it remains a product designed for AI/ML workloads (model inference, notebooks, training jobs), not a real general-purpose CaaS. A concrete limitation I've run into: outside of the GRA region, there's no CPU-only flavor with a large amount of CPU/RAM — the non-GPU flavors available in other regions are too limited for CPU-intensive container workloads. This isn't just an inconvenience: some of my projects must be hosted in Canada, so relying on GRA isn't an option for them, and there's currently no viable AI Deploy alternative in that region.

This is a real constraint for us in Canada (BHS), where data residency is required for compliance with Quebec's Law 25. Since AI Deploy doesn't offer suitable CPU/RAM-heavy, GPU-less flavors in that region, we can't use this workaround for workloads that must stay in Canada, and end up with no viable serverless container option there at all in OVHcloud.

So I want to stress that the original need is still relevant: a generic serverless container offering (similar to Cloud Run / AWS Fargate / Azure Container Apps), available consistently across all regions — including BHS for Canadian data residency requirements — would provide a much more complete answer than this AI Deploy workaround.

**Comment by Moumouls** [comment]
@antonin-a give us some infra primitives like:
- CaaS/FaaS (Google Cloud Run)
- Managed Tasks (Google Cloud Tasks)
- Object Storage (S3)
- Some IAM stuff (service accounts, groups of service accounts)
- The ability to create all these resources with an IaC tool like Terraform

For me, with these 5 basic elements well integrated, OVH could be a real solution for deploying some complex apps.
